### ISSUE TITLE
[Docs, Bug] LCM Working Properly?

### ISSUE THREAD

**Issue by jimmyrisk** [cornellius-gp/gpytorch#1553]
I'm having issues understanding the implementation of LCM in pytorch.

You can look at these slides, which are consistent with the literature: [http://gpss.cc/gpss17/slides/multipleOutputGPs.pdf](http://gpss.cc/gpss17/slides/multipleOutputGPs.pdf)

If you go to slide 29, you can see that there are D outputs, each of which is a linear combination of u_q^i, where the u_q^i are the latent GP's.

As an example, if we have D=4, Q=2, R_1 = 1 = R_2 = 1, there are 4 tasks and 2 latent GP's.  Thus, in terms of hyperparameters, I expect:

- two sets of kernel hyperparameters (one for each latent GP)
- eight parameters (D*(R_1 + R_2)) corresponding to the coefficients to reconstruct the output from the latent GP's

GPyTorch also has an added ability to add a covariance structure among the noises - this is fine and I understand it.

What I don't understand is when I fit the setup described above, I get these four objects

```
('covar_module.covar_module_list.0.task_covar_module.raw_var', tensor([-4.9917, -3.3054, -2.6581, -1.2737], device='cuda:0'))
('covar_module.covar_module_list.0.task_covar_module.covar_factor', tensor([[0.8553],
        [1.2499],
        [1.6728],
        [2.2266]], device='cuda:0'))
('covar_module.covar_module_list.1.task_covar_module.raw_var', tensor([-3.1660, -1.7249, -3.4677, -0.7606], device='cuda:0'))
('covar_module.covar_module_list.1.task_covar_module.covar_factor', tensor([[1.6271],
        [1.3652],
        [1.1850],
        [1.0030]], device='cuda:0'))
```

I'm assuming that one of raw_var or covar_factor corresponds to the linear combinations, but I'm at a loss as to what the other element corresponds to.  The lengthscales are present:

```
('covar_module.covar_module_list.0.data_covar_module.raw_lengthscale', tensor([[4.5316, 4.5134]], device='cuda:0'))
('covar_module.covar_module_list.1.data_covar_module.raw_lengthscale', tensor([[2.4409, 3.4740]], device='cuda:0'))
```

There are no scale factors present, so I'm wondering if one of raw_var or covar_factor corresponds to a scale factor, but this makes no sense, since there are four of them per kernel (and there should only be one per kernel...)

One other confusing aspect is that ``rank`` needs to be specified in ``LCMKernel`` and it also needs to be specified in ``MultitaskGaussianLikelihood``, but they are referring to different things (in ``LCMKernel``, it appears to be the number of times that each kernel repeats additively, whereas in ``MultitaskGaussianLikelihood``, it corresponds to the number of noise terms)

There is also a feature unavailable (as far as I can tell), which is that we should be able to specify multiple ranks per kernel in the list (referring to the slides, I'm saying we should be able to specify R_1, R_2, ..., R_q separately, but as far as I can tell right now, we have to set all of R_1 = R_2 = ... = R_q = ``rank``)

Here is my code for reference.  The input data is two-dimensional:

```
latent_kernels = [RBFKernel(ard_num_dims=2), RBFKernel(ard_num_dims=2)],  # 2 latent kernels
num_tasks = D  # 4, in this case
rank = 1 
rankk = 0  # specifying a different value for MultitaskGaussianLikelihood


class MultitaskGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(MultitaskGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.MultitaskMean(
            gpytorch.means.ZeroMean(), num_tasks=num_tasks
        )
        self.covar_module = gpytorch.kernels.LCMKernel(
            latent_kernels, num_tasks=num_tasks, rank=rank
        )

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultitaskMultivariateNormal(mean_x, covar_x)


likelihood = gpytorch.likelihoods.MultitaskGaussianLikelihood(num_tasks=num_tasks, rank=rankk,
                                                              noise_constraint=gpytorch.constraints.GreaterThan(1e-3)).cuda()
model = MultitaskGPModel(train_x, train_y_Stacked, likelihood).cuda()
```

This is my first time posting a message like this, so please let me know if you need any info or if I did something wrong!

edit: I also wasn't sure what to add the tag to be, since it's related to documentation and possibly a bug too, and it's also a question

**Comment by jimmyrisk** [comment]
Hi, I've made a reproducible example, essentially mimicing the code here: [https://docs.gpytorch.ai/en/stable/examples/03_Multitask_Exact_GPs/Multitask_GP_Regression.html](https://docs.gpytorch.ai/en/stable/examples/03_Multitask_Exact_GPs/Multitask_GP_Regression.html)

If LCM is used, with this code I'd expect:

- 2 GP's being fit, each with their own kernel and hyperparameter
- 4 GP's in the output, constructed as linear combination of the 2 GP's fit

Thus, I need 

- **two** total lengthscales,
- **two** total scale factors for the GP / kernel, 
- **eight** coefficients (for reconstructing the true output through the latent GP's)
- any additional parameters through noise terms

This is what I end up getting:

```
print(model.state_dict())
OrderedDict([('likelihood.raw_task_noises', tensor([-4.1334, -3.8428, -4.1504, -4.3181])), ('likelihood.raw_noise', tensor([-4.1258])), ('likelihood.raw_task_noises_constraint.lower_bound', tensor(1.0000e-04)), ('likelihood.raw_task_noises_constraint.upper_bound', tensor(inf)), ('likelihood.raw_noise_constraint.lower_bound', tensor(1.0000e-04)), ('likelihood.raw_noise_constraint.upper_bound', tensor(inf)), ('mean_module.base_means.0.constant', tensor([0.0042])), ('mean_module.base_means.1.constant', tensor([0.4178])), ('mean_module.base_means.2.constant', tensor([-0.3810])), ('mean_module.base_means.3.constant', tensor([-0.1818])), ('covar_module.covar_module_list.0.task_covar_module.covar_factor', tensor([[0.5120],
        [0.3477],
        [0.0882],
        [0.1771]])), ('covar_module.covar_module_list.0.task_covar_module.raw_var', tensor([-0.9962, -0.8105, -5.7545, -4.6294])), ('covar_module.covar_module_list.0.task_covar_module.raw_var_constraint.lower_bound', tensor(0.)), ('covar_module.covar_module_list.0.task_covar_module.raw_var_constraint.upper_bound', tensor(inf)), ('covar_module.covar_module_list.0.data_covar_module.raw_lengthscale', tensor([[-1.1819]])), ('covar_module.covar_module_list.0.data_covar_module.raw_lengthscale_constraint.lower_bound', tensor(0.)), ('covar_module.covar_module_list.0.data_covar_module.raw_lengthscale_constraint.upper_bound', tensor(inf)), ('covar_module.covar_module_list.1.task_covar_module.covar_factor', tensor([[ 0.1450],
        [ 0.6981],
        [-0.4472],
        [ 0.2070]])), ('covar_module.covar_module_list.1.task_covar_module.raw_var', tensor([-2.1072, -2.3508, -2.8507, -4.4872])), ('covar_module.covar_module_list.1.task_covar_module.raw_var_constraint.lower_bound', tensor(0.)), ('covar_module.covar_module_list.1.task_covar_module.raw_var_constraint.upper_bound', tensor(inf)), ('covar_module.covar_module_list.1.data_covar_module.raw_lengthscale', tensor([[-0.5483]])), ('covar_module.covar_module_list.1.data_covar_module.raw_lengthscale_constraint.lower_bound', tensor(0.)), ('covar_module.covar_module_list.1.data_covar_module.raw_lengthscale_constraint.upper_bound', tensor(inf))])
```

There are two lengthscales indeed, and I'm sure that one set of those 4 length vectors corresponds to coefficients, but there are two sets of four (``covar_factor`` and ``raw_var``).  

I'm also getting **four** base means.  I understand that this is set with ``num_tasks=4`` in ``mean_module``, but I believe this is another inconsistency with the LCM literature (shouldn't there be two means, one for each latent GP?)

Here is the full code for this reproducible example, with some code to plot to help verify it's at least giving the correct dimensional output.

```
import math
import torch
import gpytorch
from matplotlib import pyplot as plt

train_x = torch.linspace(0, 1, 100)

train_y = torch.stack([
    torch.sin(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2,
    torch.cos(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2,
    train_x * torch.cos(train_x * (math.pi)) + torch.randn(train_x.size()) * 0.2,
    -train_x * torch.sin(train_x * (math.pi)) + torch.randn(train_x.size()) * 0.2,
], -1)

class MultitaskGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(MultitaskGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.MultitaskMean(
            gpytorch.means.ConstantMean(), num_tasks=4
        )
        self.covar_module = gpytorch.kernels.LCMKernel(
            [gpytorch.kernels.RBFKernel(),  # 2 latent GP's
             gpytorch.kernels.MaternKernel()],
            num_tasks=4, rank=1
        )

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultitaskMultivariateNormal(mean_x, covar_x)


likelihood = gpytorch.likelihoods.MultitaskGaussianLikelihood(num_tasks=4)
model = MultitaskGPModel(train_x, train_y, likelihood)



training_iterations = 50

# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam(model.parameters(), lr=0.1)  # Includes GaussianLikelihood parameters

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

for i in range(training_iterations):
    optimizer.zero_grad()
    output = model(train_x)
    loss = -mll(output, train_y)
    loss.backward()
    print('Iter %d/%d - Loss: %.3f' % (i + 1, training_iterations, loss.item()))
    optimizer.step()

model.eval()
likelihood.eval()


# Initialize plots
f, ((y1_ax, y2_ax), (y3_ax, y4_ax)) = plt.subplots(2, 2, figsize=(8, 6))

# Make predictions
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    test_x = torch.linspace(0, 1, 51)
    predictions = likelihood(model(test_x))
    mean = predictions.mean
    lower, upper = predictions.confidence_region()

# This contains predictions for both tasks, flattened out
# The first half of the predictions is for the first task
# The second half is for the second task

# Plot training data as black stars
y1_ax.plot(train_x.detach().numpy(), train_y[:, 0].detach().numpy(), 'k*')
# Predictive mean as blue line
y1_ax.plot(test_x.numpy(), mean[:, 0].numpy(), 'b')
# Shade in confidence
y1_ax.fill_between(test_x.numpy(), lower[:, 0].numpy(), upper[:, 0].numpy(), alpha=0.5)
y1_ax.set_ylim([-3, 3])
y1_ax.legend(['Observed Data', 'Mean', 'Confidence'])

# Plot training data as black stars
y2_ax.plot(train_x.detach().numpy(), train_y[:, 1].detach().numpy(), 'k*')
# Predictive mean as blue line
y2_ax.plot(test_x.numpy(), mean[:, 1].numpy(), 'b')
# Shade in confidence
y2_ax.fill_between(test_x.numpy(), lower[:, 1].numpy(), upper[:, 1].numpy(), alpha=0.5)
y2_ax.set_ylim([-3, 3])

# Plot training data as black stars
y3_ax.plot(train_x.detach().numpy(), train_y[:, 2].detach().numpy(), 'k*')
# Predictive mean as blue line
y3_ax.plot(test_x.numpy(), mean[:, 2].numpy(), 'b')
# Shade in confidence
y3_ax.fill_between(test_x.numpy(), lower[:, 2].numpy(), upper[:, 2].numpy(), alpha=0.5)
y3_ax.set_ylim([-3, 3])

# Plot training data as black stars
y4_ax.plot(train_x.detach().numpy(), train_y[:, 3].detach().numpy(), 'k*')
# Predictive mean as blue line
y4_ax.plot(test_x.numpy(), mean[:, 3].numpy(), 'b')
# Shade in confidence
y4_ax.fill_between(test_x.numpy(), lower[:, 1].numpy(), upper[:, 3].numpy(), alpha=0.5)
y4_ax.set_ylim([-3, 3])

f.show()

print(model.state_dict())
```

**Comment by wjmaddox** [comment]
Good question, hopefully this clarifies somewhat:

> There are two lengthscales indeed, and I'm sure that one set of those 4 length vectors corresponds to coefficients, but there are two sets of four (covar_factor and raw_var).

The `covar_factor` (C) and `raw_var` (D) terms parameterize the index kernel (in the multitask kernel) as `B_q = C_q C_q + D_q` which is low rank plus diagonal. The rank setting parameterizes the rank of `C_q`. This can be seen internally [here](https://github.com/cornellius-gp/gpytorch/blob/cc717f9fbddb1d4a4e868f2411a66502ce1b1a32/gpytorch/kernels/index_kernel.py#L74) as `LMCKernel` calls `MultitaskKernel` internally which then calls `IndexKernel`. 

> There is also a feature unavailable (as far as I can tell), which is that we should be able to specify multiple ranks per kernel in the list (referring to the slides, I'm saying we should be able to specify R_1, R_2, ..., R_q separately, but as far as I can tell right now, we have to set all of R_1 = R_2 = ... = R_q = rank)

Good point, this is definitely a limitation of the current implementation, and I can try to put up a pull request. In the mean time, something like this will upgrade the rank (to modify the first covar factor to have a rank of 2, rather than 1):
```python
model.covar_module.covar_module_list[0].task_covar_module.covar_factor = torch.nn.Parameter(
    torch.randn(4, 2), requires_grad = True
)
```

> I'm also getting four base means. I understand that this is set with num_tasks=4 in mean_module, but I believe this is another inconsistency with the LCM literature (shouldn't there be two means, one for each latent GP?)

I might be wrong on this, but referring to slide 28, the LCM actually is defined as having zero mean for each latent, which should be equivalent to four zero means. Implementing two constant means and then expanding them out via the task_covariance_matrices would be tricky to do, but might be possible.

**Comment by jimmyrisk** [comment]
> The `covar_factor` (C_q) and `raw_var` (D_q) terms parameterize the index kernel (in the multitask kernel) as `B_q = C_q C_q^\top + D_q` which is low rank plus diagonal. The rank setting parameterizes the rank of `C_q`. This can be seen internally [here](https://github.com/cornellius-gp/gpytorch/blob/cc717f9fbddb1d4a4e868f2411a66502ce1b1a32/gpytorch/kernels/index_kernel.py#L74) as `LMCKernel` calls `MultitaskKernel` internally which then calls `IndexKernel`.
> 

Thank you!  This is the exact clarification I was looking for.  I think I can back out the needed coefficients from this representation.


> > I'm also getting four base means. I understand that this is set with num_tasks=4 in mean_module, but I believe this is another inconsistency with the LCM literature (shouldn't there be two means, one for each latent GP?)
> 
> I might be wrong on this, but referring to slide 28, the LCM actually is defined as having zero mean for each latent, which should be equivalent to four zero means. Implementing two constant means and then expanding them out via the task_covariance_matrices would be tricky to do, but might be possible.

For reference, here is a well cited paper that discusses it more in detail: [https://arxiv.org/pdf/1106.6251.pdf](https://arxiv.org/pdf/1106.6251.pdf)

I am confused, because I was thinking "with zero mean" was more of a "without loss of generality, it has zero mean" type thing.  On slide 28 of the slides, as well as p12 of the pdf, we can see that the multi-task GP is defined explicitly through latent GP's.  Since the latent GP's have zero prior mean, it implies that each f has zero prior mean as well.

Thus, if we wanted to add a mean, I suppose we could just consider f_d(x) + mu_d(x), and have mu_d(x) fitted from the data (this seems to be what GPyTorch is doing now).  

I thought it would be more consistent with the overall theme of LCM to have it such that each latent GP has its own prior mean, where by "theme" I mean that the multi-task GP is completely defined as linear combinations of the latent GP's, though I could be wrong.  Regardless, I think now that I understand the current implementation of GPyTorch, I can make it work with my project.

I think it's important to be consistent with how the literature is presenting LCM, however, if it's consistent throughout the literature.  For example, I'm now interested in writing up a doc to help explain LCM in GPyTorch, but I feel like the presentation may be inconsistent with the literature.

Side note, but related:  I think "linear model of coregionalization" (LMC) and "linear coregionalization model" (LCM) mean the same thing, but I haven't seen "LCM" aside from the paper referenced in this [doc from GPyTorch](https://docs.gpytorch.ai/en/stable/examples/03_Multitask_Exact_GPs/index.html) ([paper link](https://homepages.inf.ed.ac.uk/ckiw/postscript/multitaskGP_v22.pdf)).  A related issue brought up on github for this package is [here](https://github.com/cornellius-gp/gpytorch/issues/700), but it never seemed to be fully resolved.  Could we get some clarity about it?

**Comment by wjmaddox** [comment]
> I am confused, because I was thinking "with zero mean" was more of a "without loss of generality, it has zero mean" type thing. On slide 28 of the slides, as well as p12 of the pdf, we can see that the multi-task GP is defined explicitly through latent GP's. Since the latent GP's have zero prior mean, it implies that each f has zero prior mean as well.

Yes, that makes sense. I worked out the math afterwards and it does look like it's fine to have non zero means in the model specification of the LMC. Implementation wise, that could be a bit tricky to code up. Somewhat relatedly, I was actually struggling to find another reference implementation of the LMC to see how the non-zero prior mean is typically implemented, if you know of one that would be great.

> I think it's important to be consistent with how the literature is presenting LCM, however, if it's consistent throughout the literature. For example, I'm now interested in writing up a doc to help explain LCM in GPyTorch, but I feel like the presentation may be inconsistent with the literature.

cc @Balandat @jacobrgardner who implemented the LCM and would likely know a bit more.

> Side note, but related: I think "linear model of coregionalization" (LMC) and "linear coregionalization model" (LCM) mean the same thing, but I haven't seen "LCM" aside from the paper referenced in this doc from GPyTorch (paper link). A related issue brought up on github for this package is here, but it never seemed to be fully resolved. Could we get some clarity about it?

Best that I can tell, they seem to be completely interchangeable in the literature. Searching "linear coregionalization model" on google scholar finds https://link.springer.com/chapter/10.1007/978-94-007-4153-9_6 and https://link.springer.com/article/10.1007/BF00893750 as the top hits, where it's pretty clearly interchangeable.

**Comment by gpleiss** [comment]
The LCM implementation is stable. Closing this issue for now.
